src/rewrites.rs

```diff
@@ -588,3 +712,274 @@ fn check_pat(
+
+                                // check_pat on both dst patterns
+                                if check_pat(rule.2.ast.as_ref(), &mut runner.egraph, &merged_subst)
+                                    .0
+                                {
+                                    if check_pat(
+                                        rule.3.ast.as_ref(),
+                                        &mut runner.egraph,
+                                        &merged_subst,
+                                    )
+                                    .0
+                                    {
+                                        // apply dst patterns, union
+                                        let id_1 = rule.2.apply_one(
+                                            &mut runner.egraph,
+                                            match_1.eclass,
+                                            &merged_subst,
+                                        )[0];
+                                        runner.egraph.union(id_1, match_1.eclass);
+                                        let id_2 = rule.3.apply_one(
+                                            &mut runner.egraph,
+                                            match_2.eclass,
+                                            &merged_subst,
+                                        )[0];
+                                        runner.egraph.union(id_2, match_2.eclass);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        runner.egraph.rebuild();
+        Ok(())
+    }
+}
+
+/// Canonicalize a pattern
+///
+/// This function constructs a canonicalized pattern for a given pattern. It sequentially
+/// replaces each unique variable with ?i_0, ?i_1 ...
+///
+/// # Parameters
+///
+/// - `pat`: the pattern to canonicalize. See egg::Pattern for more info
+///
+/// # Returns
+///
+/// - Pattern<Mdl>: the canonicalized pattern
+/// - HashMap<egg::Var, egg::Var>: Mapping from variable in the original pattern to
+///     variable in the canonical pattern.
+fn canonicalize(pat: &Pattern<Mdl>) -> (Pattern<Mdl>, HashMap<egg::Var, egg::Var>) {
+    let mut var_map = HashMap::<egg::Var, egg::Var>::new();
+    let mut count = 0;
+    let substituted: Vec<_> = pat
+        .ast
+        .as_ref()
+        .iter()
+        .cloned()
+        .map(|x| match x {
+            ENodeOrVar::ENode(_) => x,
+            ENodeOrVar::Var(v) => match var_map.get(&v) {
+                Some(var) => ENodeOrVar::Var(*var),
+                None => {
+                    let name = format!("?i_{}", count);
+                    count += 1;
+                    let new_var: egg::Var = name.parse().unwrap();
+                    let new_node = ENodeOrVar::Var(new_var);
+                    var_map.insert(v, new_var);
+                    new_node
```

Comment:
Can't you write this as
```
var_map.insert(v, new_var);
ENodeOrVar::Var(new_var);
```

If so, you can also rewrite the whole match as
```
let var = var_map.get(&v).or_else(|| {
  let name = format!("?i_{}", count);
  count += 1;
  let new_var: egg::Var = name.parse().unwrap();
  var_map.insert(v, new_var);
  new_var
});
ENodeOrVar::Var(var)
```

Reply:
Done

Reply:
Sorry, didn't realize you where inserting in the same hashmap
```
let var = var_map.entry(v).or_insert_with(|| {
  let name = format!("?i_{}", count);
  count += 1;
  name.parse().unwrap()
})
```

Reply:
Done